File: peak.java
//TC:O(n)
//SC:O(n)
//solved in leetcode:YES
class Solution {
    public int findPeakElement(int[] nums) {
        int low=0;
        int high = nums.length-1;
        System.out.println(high);
        while(low<=high)
        {
            int mid = low+(high-low)/2;
            System.out.println(mid);
            // If the element is first or the only element, return, else compare with the neighbours
            if((mid==0||nums[mid]>nums[mid-1]) &&((mid==nums.length-1 || nums[mid]>nums[mid+1])))
            {
                System.out.println(mid);
                return mid;

            }
            // go to the highest side while trying to get peak
            if(nums[mid]<nums[mid+1])
            {
                low=mid+1;
            }
            else
            {
                high=mid-1;
            }


        }
        return -1;
    }
}

File: peak.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int n = nums.size();
        if(n==1){return 0;}
        int low = 0, high = nums.size()-1;
        int mid ;
        while(low<=high){
            mid = low+ (high-low)/2;
            if(mid>0 && mid<n-1){
                if((nums[mid-1]<nums[mid]) && (nums[mid+1]<nums[mid])){
                    return mid;
                }
                else if((nums[mid-1]>nums[mid])){
                    high = mid-1;
                }
                else{
                    low = mid+1;
                }
            }
          else if(mid==0){
              if(nums[mid]>nums[mid+1]){return mid;}
              else{return mid+1;}
          }
          else if(mid==n-1){
              if(nums[mid]>nums[mid-1]){return mid;}
              else{return mid-1;}
          }

        }
    return -1;
    }
};
